File: iconanim/src/main/java/sgtmelon/iconanim/app/SwitchButton.java
package sgtmelon.iconanim.app;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.TypedValue;

import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.core.content.ContextCompat;
import sgtmelon.iconanim.R;
import sgtmelon.iconanim.office.intf.AnimIntf;

public class SwitchButton extends AppCompatImageButton implements AnimIntf {

    final Context context;

    @DrawableRes
    protected int srcDisable, srcSelect;
    @DrawableRes
    protected int srcDisableAnim, srcSelectAnim;
    @ColorInt
    protected int srcDisableColor, srcSelectColor;

    Drawable drawableDisable, drawableSelect;

    public SwitchButton(Context context, AttributeSet attrs) {
        super(context, attrs);

        this.context = context;

        initAttribute(attrs);

        setupDrawable();
    }

    public SwitchButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        this.context = context;

        initAttribute(attrs);

        setupDrawable();
    }

    private void initAttribute(AttributeSet attributeSet) {
        TypedArray attributes = context.obtainStyledAttributes(attributeSet, R.styleable.SwitchButton);

        srcDisable = attributes.getResourceId(R.styleable.SwitchButton_srcDisable, R.drawable.ic_android);
        srcSelect = attributes.getResourceId(R.styleable.SwitchButton_srcSelect, R.drawable.ic_android);

        srcDisableAnim = attributes.getResourceId(R.styleable.SwitchButton_srcDisableAnim, R.drawable.ic_android);
        srcSelectAnim = attributes.getResourceId(R.styleable.SwitchButton_srcSelectAnim, R.drawable.ic_android);

        TypedValue typedValue = new TypedValue();
        context.getTheme().resolveAttribute(R.attr.srcDisableColor, typedValue, true);
        srcDisableColor = ContextCompat.getColor(context, typedValue.resourceId);

        context.getTheme().resolveAttribute(R.attr.srcSelectColor, typedValue, true);
        srcSelectColor = ContextCompat.getColor(context, typedValue.resourceId);

        attributes.recycle();
    }

    void setupDrawable() {
        drawableDisable = ContextCompat.getDrawable(context, srcDisable);
        if (drawableDisable != null) {
            drawableDisable.setColorFilter(srcDisableColor, PorterDuff.Mode.SRC_ATOP);
        }

        drawableSelect = ContextCompat.getDrawable(context, srcSelect);
        if (drawableSelect != null) {
            drawableSelect.setColorFilter(srcSelectColor, PorterDuff.Mode.SRC_ATOP);
        }

        setImageDrawable(drawableSelect);
    }

    @Override
    public void setDrawable(boolean select, boolean needAnim) {
        if (select) setImageDrawable(drawableSelect);
        else setImageDrawable(drawableDisable);
    }

}

File: libraries/iconAnim/src/main/java/sgtmelon/iconanim/widget/SwitchButton.kt
package sgtmelon.iconanim.widget

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.AnimatedVectorDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.AppCompatImageButton
import sgtmelon.extensions.getDrawableCompat
import sgtmelon.extensions.setColorFilterCompat
import sgtmelon.iconanim.R
import sgtmelon.iconanim.callback.IconBlockCallback
import sgtmelon.iconanim.callback.ParentIconChange
import sgtmelon.iconanim.control.AnimatedIcon

/**
 * Button with automatic icon change via [setDrawable] func.
 */
class SwitchButton(
    context: Context,
    attrs: AttributeSet
) : AppCompatImageButton(context, attrs),
    ParentIconChange {

    @DrawableRes private val srcSelect: Int
    @DrawableRes private val srcDisable: Int
    @DrawableRes private val srcSelectAnim: Int
    @DrawableRes private val srcDisableAnim: Int

    @ColorInt private val srcSelectColor: Int
    @ColorInt private val srcDisableColor: Int

    private val iconSelect: Drawable?
    private val iconDisable: Drawable?
    private val iconSelectAnim: AnimatedVectorDrawable?
    private val iconDisableAnim: AnimatedVectorDrawable?

    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.SwitchButton)

        srcSelect = array.getResourceId(R.styleable.SwitchButton_srcSelect, ND_SRC)
        srcDisable = array.getResourceId(R.styleable.SwitchButton_srcDisable, ND_SRC)

        srcSelectAnim = array.getResourceId(R.styleable.SwitchButton_srcSelectAnim, ND_SRC)
        srcDisableAnim = array.getResourceId(R.styleable.SwitchButton_srcDisableAnim, ND_SRC)

        srcSelectColor = array.getColor(R.styleable.SwitchButton_srcSelectColor, Color.BLACK)
        srcDisableColor = array.getColor(R.styleable.SwitchButton_srcDisableColor, Color.BLACK)

        array.recycle()

        iconSelect = if (srcSelect != ND_SRC) context.getDrawableCompat(srcSelect) else null
        iconDisable = if (srcDisable != ND_SRC) context.getDrawableCompat(srcDisable) else null

        iconSelectAnim = if (srcSelectAnim != ND_SRC) {
            context.getDrawableCompat(srcSelectAnim) as? AnimatedVectorDrawable
        } else {
            null
        }
        iconDisableAnim = if (srcDisableAnim != ND_SRC) {
            context.getDrawableCompat(srcDisableAnim) as? AnimatedVectorDrawable
        } else {
            null
        }
    }

    init {
        iconSelect?.setColorFilterCompat(srcSelectColor)
        iconDisable?.setColorFilterCompat(srcDisableColor)
        iconSelectAnim?.setColorFilterCompat(srcSelectColor)
        iconDisableAnim?.setColorFilterCompat(srcDisableColor)
    }

    override val animatedIcon = AnimatedIcon(
        context, iconSelectAnim, iconDisableAnim, changeCallback = this
    )

    fun setBlockCallback(blockCallback: IconBlockCallback) {
        animatedIcon.setBlockCallback(blockCallback)
    }

    override val enterIcon: Drawable? = iconSelect
    override val exitIcon: Drawable? = iconDisable

    override var isEnterIcon: Boolean? = null

    override fun setDrawableAfterChange(drawable: Drawable?) = setImageDrawable(drawable)

    companion object {
        const val ND_SRC = -1
    }
}